src/pipx/pipx_metadata_file.py

```diff
@@ -0,0 +1,158 @@
+        #   members, EXCEPT for those we cannot know:
+        #       self.main_package.package_or_url=None
+        #       self.venv_metadata.package_or_url=None
+        self.main_package = PackageInfo(
+            package=None,
+            package_or_url=None,
+            pip_args=[],
+            include_dependencies=False,
+            include_apps=True,  # always True for main_package
+            apps=[],
+            app_paths=[],
+            apps_of_dependencies=[],
+            app_paths_of_dependencies={},
+            package_version="",
+        )
+        self.python_version = None
+        self.venv_args = []
+        self.injected_packages = {}
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "main_package": self.main_package._asdict(),
+            "python_version": self.python_version,
+            "venv_args": self.venv_args,
+            "injected_packages": {
+                name: data._asdict() for (name, data) in self.injected_packages.items()
+            },
+            "pipx_metadata_version": self._pipx_metadata_version,
+        }
+
+    def from_dict(self, input_dict: Dict[str, Any]) -> None:
+        self.main_package = PackageInfo(**input_dict["main_package"])
+        self.python_version = input_dict["python_version"]
+        self.venv_args = input_dict["venv_args"]
+        self.injected_packages = {
+            name: PackageInfo(**data)
+            for (name, data) in input_dict["injected_packages"].items()
+        }
+
+    def _validate_before_write(self):
+        if (
+            self.main_package.package is None
+            or self.main_package.package_or_url is None
+            or not self.main_package.include_apps
+        ):
+            raise PipxError("Internal Error: PipxMetadata is corrupt, cannot write.")
+
+    def write(self) -> None:
+        self._validate_before_write()
+        try:
+            with open(self.venv_dir / PIPX_INFO_FILENAME, "w") as pipx_metadata_fh:
+                json.dump(
+                    self.to_dict(),
+                    pipx_metadata_fh,
+                    indent=4,
+                    sort_keys=True,
+                    cls=JsonEncoderHandlesPath,
+                )
+        except IOError:
+            logging.warning(
+                textwrap.fill(
+                    f"Unable to write {PIPX_INFO_FILENAME} to {self.venv_dir}. "
+                    f"This may cause future pipx operations involving "
+                    f"{self.venv_dir.name} to fail or behave incorrectly.",
+                    width=79,
+                )
+            )
+            pass
+
+    def read(self, verbose: bool = False) -> None:
+        try:
+            with open(self.venv_dir / PIPX_INFO_FILENAME, "r") as pipx_metadata_fh:
+                self.from_dict(
+                    json.load(pipx_metadata_fh, object_hook=_json_decoder_object_hook)
+                )
```

Comment:
I always try to reduce the number of side effects and make functions pure since the code is almost always easier to follow. In this case, I would suggest returning 
```
json.load(pipx_metadata_fh, object_hook=_json_decoder_object_hook)
```
rather than having the side effect of calling `from_dict`, then in the caller call `from_dict`.

Reply:
I had a different idea of the purpose of this.  To my mind `read()`'s main purpose *was* to set all of the fields of `PipxMetadataFile` from file data.  To my mind what's going on here are implementation details of the main effect and not a side effect.

Reply:
I totally understand the idea of having no side effects and how it is beneficial.  Maybe you can help me understand how it applies here.  

Basically, my intention was that the user of this python module never has to deal with file access or see raw dicts from the file access.  I was trying to wrap that completely in the object.  So I didn't make a function that would return the raw dict itself, and tried to make the code so you would never need to get or manipulate the raw dict.

Is there a way to structure this code well, but still take care of everything internal to the object like I was envisioning?

Reply:
I think it's fine as is.